### DISCUSSION TITLE
Abstraction of dual user object types through a common base class

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#19485] [Q&A General] [answered]
Hi,
I am trying to develop a new "transfer" class which directly push the user object variables on to the multi-apps. Here is the block i am interested in.
<code><pre>
[UserObjects]
  [stress_tensor]
    type = StressTensorUO
    stress_variable = 'stress_xx_av'
  []
  [def_coords]
    type = DefCoordsUO
    stress_variable = 'stress_xx_av'
  []
[]

[Transfers]
  [transfer_qty_fwd]
    type = MultiAppTransferQuantity_UO
    direction = to_multiapp
    user_objects_name = "stress_tensor def_coords"
    user_object_types = "StressTensorUO DefCoordsUO"
    subapp_index = 0
    multi_app = sub
  []
[]
</code></pre>

One common issue being that, how to create an user object variable that can open up access to data inside both `stress_tensor ` and `def_coords` through an universal type. In order to address that, i created a common base class "MultiAppUserObjBase" and asked my user object classes to derive from this base class. When i try to access the user object data through this base class object (as given below), from my transfer class, i am getting undefined reference, which is quite understandable. 
<code><pre>
FEProblemBase &  problem =  *_from_problems[0];
  _transfer_uo[0] = &problem.getUserObject<MultiAppUserObjBase>(_uo_names[0]);
  for (auto i:_transfer_uo[0]->_stress_vec){
    printf("New Stress Vector: %f",i)
  }
</code></pre>

<code><pre>
/home/abalasub/Working/Projects/moose/framework/src/transfers/MultiAppTransferQuantity_UO.C: In member function ‘virtual void MultiAppTransferQuantity_UO::executeToMultiapp()’:
/home/abalasub/Working/Projects/moose/framework/src/transfers/MultiAppTransferQuantity_UO.C:80:32: error: ‘const class MultiAppUserObjBase’ has no member named ‘_stress_vec’
   for (auto i:_transfer_uo[0]->_stress_vec){
                                ^~~~~~~~~~~
make: *** [/home/abalasub/Working/Projects/moose/framework/build/unity_src/transfers_Unity.x86_64-pc-linux-gnu.opt.lo] Error 1
</code></pre>

Possibly i am missing something here, may be an upcasting to the child objects, which i am not very sure. The attached files might give you idea of the developments that has happened in this work. Your inputs here will be highly appreciated.
Kind regards,
Arun

**Comment by abarun22** [comment]
Sorry about that. Here are the attachments missing from my previous post.

[25112021.zip](https://github.com/idaholab/moose/files/7603363/25112021.zip)

**Comment by GiudGiud** [comment, marked as answer]
Hello

What do you mean by UO variables? Are they auxiliary variables? These can be transfered already. 

I'd be a little wary of inheriting from two unrelated systems (UOs and transfers). It often works, but you have to be careful to consider all the interactions. 

I think the cleaner code would use composition instead of inheritance. The transfer would have a UO parameter, and would access the things it needs to transfer from the UO using accessor to the UO's member variables. 

Guillaume
